File: data/data_file.go
package data

import (
	"errors"
	"fmt"
	"hash/crc32"
	"io"
	"path/filepath"

	"github.com/yifaaan/bitcask/fio"
)

const (
	DATA_FILE_NAME_SUFFIX    = ".data"
	HINT_FILE_NAME           = "hint-index"
	MERGE_FINISHED_FILE_NAME = "merge-finished"
)

var (
	ErrInvalidCRC = errors.New("invalid crc value")
)

// DataFile 数据文件
type DataFile struct {
	FileId    uint32
	WriteOff  int64
	IOManager fio.IOManager
}

func OpenDataFile(dirPath string, fid uint32, ioType fio.FileIOType) (*DataFile, error) {
	name := filepath.Join(dirPath, fmt.Sprintf("%09d", fid)+DATA_FILE_NAME_SUFFIX)
	return newDataFile(name, fid, ioType)
}

func OpenHintFile(dirPath string) (*DataFile, error) {
	name := filepath.Join(dirPath, HINT_FILE_NAME)
	return newDataFile(name, 0, fio.STANDARD_FIO)
}

func OpenMergeFinishedFile(dirPath string) (*DataFile, error) {
	name := filepath.Join(dirPath, MERGE_FINISHED_FILE_NAME)
	return newDataFile(name, 0, fio.STANDARD_FIO)
}

func newDataFile(name string, fid uint32, ioType fio.FileIOType) (*DataFile, error) {
	iom, err := fio.NewIOManager(name, ioType)
	if err != nil {
		return nil, err
	}
	return &DataFile{
		FileId:    fid,
		WriteOff:  0,
		IOManager: iom,
	}, nil
}

func (df *DataFile) Sync() error {
	return df.IOManager.Sync()
}

func (df *DataFile) Write(buf []byte) error {
	n, err := df.IOManager.Write(buf)
	if err != nil {
		return err
	}
	df.WriteOff += int64(n)
	return nil
}

// WriteHintRecord 写入一条 hint 索引记录
func (df *DataFile) WriteHintRecord(key []byte, pos *LogRecordPos) error {
	lr := &LogRecord{
		Key:   key,
		Value: EncodeLogRecordPos(pos),
	}
	buf, _ := EncodeLogRecord(lr)
	return df.Write(buf)
}

func (df *DataFile) Close() error {
	return df.IOManager.Close()
}

// ReadLogRecord 从数据文件 off 处读取一条 record，返回其在文件中的长度
func (df *DataFile) ReadLogRecord(off int64) (*LogRecord, int64, error) {
	fileSize, err := df.IOManager.Size()
	if err != nil {
		return nil, 0, err
	}

	// 从 off 处不足以读取 MAX_LOG_RECORD_HEADER_SIZE，则需要读到文件末尾
	var headSizeToRead int64 = MAX_LOG_RECORD_HEADER_SIZE
	if off+MAX_LOG_RECORD_HEADER_SIZE > fileSize {
		headSizeToRead = fileSize - off
	}

	// header
	headerBuf, err := df.readNBytes(headSizeToRead, off)
	if err != nil {
		return nil, 0, err
	}

	header, headSize := decodeLogRecordHeader(headerBuf)

	if header == nil {
		return nil, 0, io.EOF
	}
	if header.crc == 0 && header.keySize == 0 && header.valueSize == 0 {
		return nil, 0, io.EOF
	}

	keySize, valueSize := header.keySize, header.valueSize
	// 编码后的 record 所占文件大小
	recordSize := headSize + keySize + valueSize

	lr := &LogRecord{Type: header.recordType}
	// key value
	if keySize > 0 || valueSize > 0 {
		kvBuf, err := df.readNBytes(int64(keySize)+int64(valueSize), off+int64(headSize))
		if err != nil {
			return nil, 0, err
		}
		lr.Key = kvBuf[:keySize]
		lr.Value = kvBuf[keySize:]
	}

	// 检验 crc
	if header.crc != getLogRecordCRC(lr, headerBuf[crc32.Size:headSize]) {
		return nil, 0, ErrInvalidCRC
	}
	return lr, int64(recordSize), nil
}

// readNBytes 从 off 偏移处读取 n 个字节
func (df *DataFile) readNBytes(n int64, off int64) ([]byte, error) {
	buf := make([]byte, n)
	_, err := df.IOManager.Read(buf, off)
	return buf, err
}

func (df *DataFile) SetIOManager(dirPath string, ioType fio.FileIOType) error {
	if err := df.IOManager.Close(); err != nil {
		return err
	}
	name := filepath.Join(dirPath, fmt.Sprintf("%09d", df.FileId)+DATA_FILE_NAME_SUFFIX)
	iom, err := fio.NewIOManager(name, ioType)
	if err != nil {
		return err
	}
	df.IOManager = iom
	return nil
}

File: Source/Data/DataFile.cpp
#include "DataFile.h"

#include <absl/status/status.h>

#include <algorithm>
#include <filesystem>
#include <iomanip>
#include <limits>
#include <sstream>

namespace bitcask
{

	namespace
	{

		absl::StatusOr<std::unique_ptr<DataFile>> OpenNamedFile(std::string_view dirPath, std::string_view fileName, uint32_t fid, IOType ioType)
		{
			const auto path = (std::filesystem::path(dirPath) / fileName).string();
			auto io = IOManager::NewIOManager(path, ioType);
			if (!io.ok())
			{
				return io.status();
			}

			auto file = std::make_unique<DataFile>();
			file->fid = fid;
			file->io = std::move(*io);

			auto sizeOr = file->io->Size();
			if (!sizeOr.ok())
			{
				return sizeOr.status();
			}
			file->writeOffset = *sizeOr;
			return file;
		}

		uint32_t ReadLittleEndian32(std::span<const std::byte> buf)
		{
			return static_cast<uint32_t>(std::to_integer<uint8_t>(buf[0])) |
				   (static_cast<uint32_t>(std::to_integer<uint8_t>(buf[1])) << 8) |
				   (static_cast<uint32_t>(std::to_integer<uint8_t>(buf[2])) << 16) |
				   (static_cast<uint32_t>(std::to_integer<uint8_t>(buf[3])) << 24);
		}

	} // namespace

	std::string DataFileName(uint32_t fid)
	{
		std::ostringstream stream;
		stream << std::setw(9) << std::setfill('0') << fid << DataFileNameSuffix;
		return stream.str();
	}

	absl::StatusOr<std::unique_ptr<DataFile>> bitcask::DataFile::Open(std::string_view dirPath, uint32_t fid, IOType ioType)
	{
		return OpenNamedFile(dirPath, DataFileName(fid), fid, ioType);
	}

	absl::StatusOr<std::unique_ptr<DataFile>> DataFile::OpenHint(std::string_view dirPath, uint32_t fid)
	{
		return OpenNamedFile(dirPath, HintFileName, fid, IOType::Standard);
	}

	absl::StatusOr<std::unique_ptr<DataFile>> DataFile::OpenMergeFinishedFile(std::string_view dirPath)
	{
		return OpenNamedFile(dirPath, MergeFinishedFileName, 0, IOType::Standard);
	}

	absl::Status DataFile::Sync()
	{
		return io->Sync();
	}

	absl::StatusOr<int64_t> DataFile::Write(std::span<const std::byte> data)
	{
		auto writeOr = io->Write(data);
		if (!writeOr.ok())
		{
			return writeOr.status();
		}
		if (*writeOr != data.size())
		{
			return absl::InternalError("short write");
		}

		writeOffset += *writeOr;
		return *writeOr;
	}

	absl::StatusOr<std::pair<int64_t, LogRecord>> DataFile::ReadLogRecord(int64_t offset) const
	{
		auto sizeOr = io->Size();
		if (!sizeOr.ok())
		{
			return sizeOr.status();
		}
		const int64_t fileSize = *sizeOr;
		if (offset >= fileSize)
		{
			return absl::OutOfRangeError("offset beyond file end");
		}
		const int64_t remaining = fileSize - offset;

		// Read only the prefix needed to decode the variable-length header.
		// A short read that cannot even cover the minimal header is treated as EOF/truncation.
		const auto maxHeaderRead = static_cast<size_t>(std::min<int64_t>(remaining, MaxLogRecordHeaderSize));
		std::vector<std::byte> headerBuf(maxHeaderRead);
		auto headerSizeOr = io->Read(headerBuf, offset);
		if (!headerSizeOr.ok())
		{
			return headerSizeOr.status();
		}
		headerBuf.resize(*headerSizeOr);
		if (*headerSizeOr < 5)
		{
			return absl::OutOfRangeError("unexpected eof");
		}

		auto headerOr = DecodeLogRecordHeader(headerBuf);
		if (!headerOr.first)
		{
			// If we could not read the full maximum header prefix, the record is most
			// likely cut off at the file tail. Only a fully read header prefix that
			// still cannot be decoded is considered corruption.
			if (static_cast<size_t>(*headerSizeOr) < MaxLogRecordHeaderSize)
			{
				return absl::OutOfRangeError("truncated log record header");
			}
			return absl::InternalError("invalid log record header");
		}

		const auto& header = *headerOr.first;
		if (header.crc == 0 && header.keySize == 0 && header.valueSize == 0)
		{
			return absl::OutOfRangeError("empty log record");
		}

		const auto headerSize = headerOr.second;
		const auto totalSize = headerSize + header.keySize + header.valueSize;
		if (totalSize > static_cast<uint64_t>(remaining))
		{
			return absl::OutOfRangeError("truncated log record");
		}

		// Re-read the whole record only after the header and sizes are validated.
		std::vector<std::byte> kvBuf(header.keySize + header.valueSize);
		auto kvBufSizeOr = io->Read(kvBuf, offset + headerSize);
		if (!kvBufSizeOr.ok())
		{
			return kvBufSizeOr.status();
		}
		if (*kvBufSizeOr != header.keySize + header.valueSize)
		{
			return absl::OutOfRangeError("truncated log record");
		}

		LogRecord record;
		record.type = header.type;
		// Layout: [crc(4)][type][key varint][value varint][key bytes][value bytes].
		record.key.assign(reinterpret_cast<const char*>(kvBuf.data()), header.keySize);
		record.value.assign(reinterpret_cast<const char*>(kvBuf.data() + header.keySize), header.valueSize);

		// CRC covers the header without CRC plus key/value payload.
		if (CalcLogRecordCRC(record, header) != header.crc)
		{
			return absl::InternalError("crc mismatch");
		}

		return std::make_pair(totalSize, std::move(record));
	}

	absl::Status DataFile::Close()
	{
		return io->Close();
	}

	absl::Status DataFile::WriteHintRecord(std::string_view key, const LogRecordPos& pos)
	{
		// 将 LogRecordPos 编码为字节序列, 作为 hint 记录的 value
		auto [posBytes, posSize] = EncodeLogRecordPos(pos);
		if (posSize == 0)
		{
			return absl::InvalidArgumentError("invalid LogRecordPos: offset or size is negative");
		}

		// hint 记录本质上是一条 Normal 类型的 LogRecord, value 为编码后的 LogRecordPos
		LogRecord record;
		record.key = std::string(key);
		record.value.assign(reinterpret_cast<const char*>(posBytes.data()), static_cast<size_t>(posSize));
		record.type = LogRecordType::Normal;

		auto encoded = EncodeLogRecord(record);
		if (encoded.empty())
		{
			return absl::InternalError("failed to encode hint record");
		}

		auto writeOr = Write(encoded);
		if (!writeOr.ok())
		{
			return writeOr.status();
		}
		return absl::OkStatus();
	}

	absl::StatusOr<std::pair<int64_t, HintRecord>> DataFile::ReadHintRecord(int64_t offset) const
	{
		auto recOr = ReadLogRecord(offset);
		if (!recOr.ok())
		{
			return recOr.status();
		}
		auto& [size, record] = *recOr;

		// hint 记录的 value 是编码后的 LogRecordPos, 解码必须正好消耗整个 value
		auto valueSpan = std::span<const std::byte>(
			reinterpret_cast<const std::byte*>(record.value.data()), record.value.size());
		auto [posOpt, posLen] = DecodeLogRecordPos(valueSpan);
		if (!posOpt || posLen != static_cast<int64_t>(record.value.size()))
		{
			return absl::InternalError("invalid hint record: failed to decode LogRecordPos");
		}

		HintRecord hint{
			.key = std::move(record.key),
			.pos = *posOpt,
		};
		return std::make_pair(size, std::move(hint));
	}
} // namespace bitcask
